File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/proposer/common/TheAbstractBestPriceProposerTest.java
package cz.fb.manaus.reactor.betting.proposer.common;

import cz.fb.manaus.core.model.Side;
import cz.fb.manaus.core.test.AbstractLocalTestCase;
import cz.fb.manaus.core.test.CoreTestFactory;
import cz.fb.manaus.reactor.ReactorTestFactory;
import cz.fb.manaus.reactor.betting.BetContext;
import cz.fb.manaus.reactor.betting.validator.ValidationResult;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class TheAbstractBestPriceProposerTest extends AbstractLocalTestCase {

    @Autowired
    private LayProposer layProposer;
    @Autowired
    private BackProposer backProposer;
    @Autowired
    private ReactorTestFactory factory;

    @Test
    public void testLayPropose() {
        var context = mock(BetContext.class);
        when(context.getSide()).thenReturn(Side.LAY);
        when(context.getRunnerPrices()).thenReturn(factory.newRP(CoreTestFactory.HOME, 2d, 4.5d));
        assertThat(layProposer.validate(context), is(ValidationResult.ACCEPT));
        assertThat(layProposer.getProposedPrice(context).getAsDouble(), is(2.02d));
    }

    @Test
    public void testCheck() {
        var context = mock(BetContext.class);
        when(context.getSide()).thenReturn(Side.LAY, Side.BACK);
        when(context.getRunnerPrices()).thenReturn(CoreTestFactory.newBackRP(2d, CoreTestFactory.HOME, 2d));
        assertThat(layProposer.validate(context), is(ValidationResult.ACCEPT));
        assertThat(backProposer.validate(context), is(ValidationResult.REJECT));
    }

    @Test
    public void testBackPropose() {
        var context = mock(BetContext.class);
        when(context.getSide()).thenReturn(Side.BACK);
        when(context.getRunnerPrices()).thenReturn(factory.newRP(CoreTestFactory.HOME, 2.5d, 3.5d));
        assertThat(backProposer.validate(context), is(ValidationResult.ACCEPT));
        assertThat(backProposer.getProposedPrice(context).getAsDouble(), is(3.45d));
    }

    @Component
    private static class LayProposer extends AbstractBestPriceProposer {

        public LayProposer() {
            super();
        }
    }

    @Component
    private static class BackProposer extends AbstractBestPriceProposer {

        public BackProposer() {
            super();
        }
    }

}

File: mns-reactor/src/test/java/cz/fb/manaus/reactor/betting/proposer/common/TheAbstractBestPriceProposerTest.kt
package cz.fb.manaus.reactor.betting.proposer.common

import cz.fb.manaus.core.test.AbstractLocalTestCase
import cz.fb.manaus.reactor.ReactorTestFactory
import cz.fb.manaus.reactor.betting.HOME_EVENT_BACK
import cz.fb.manaus.reactor.betting.HOME_EVENT_LAY
import cz.fb.manaus.reactor.betting.validator.ValidationResult
import org.junit.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import kotlin.test.assertEquals


class TheAbstractBestPriceProposerTest : AbstractLocalTestCase() {

    @Autowired
    private lateinit var layProposer: LayProposer
    @Autowired
    private lateinit var backProposer: BackProposer
    @Autowired
    private lateinit var factory: ReactorTestFactory

    @Test
    fun `lay propose`() {
        val event = HOME_EVENT_LAY.copy(marketPrices = factory.newMarketPrices(2.0, 4.5))
        assertEquals(ValidationResult.OK, layProposer.validate(event))
        assertEquals(2.02, layProposer.getProposedPrice(event))
    }

    @Test
    fun check() {
        assertEquals(ValidationResult.OK, layProposer.validate(HOME_EVENT_LAY))
        assertEquals(ValidationResult.OK, backProposer.validate(HOME_EVENT_BACK))
    }

    @Test
    fun `back propose`() {
        val prices = factory.newMarketPrices(2.5, 3.5)
        val context = HOME_EVENT_BACK.copy(marketPrices = prices)
        assertEquals(ValidationResult.OK, backProposer.validate(context))
        assertEquals(3.45, backProposer.getProposedPrice(context))
    }

    @Component
    private class LayProposer : AbstractBestPriceProposer(1)

    @Component
    private class BackProposer : AbstractBestPriceProposer(1)

}
